File: python/luhn/luhn.py
import re


def doubling(number):
    double = number * 2
    if double < 9:
        return double

    return double - 9


class Luhn:
    def __init__(self, card_num):
        self.sequence = card_num

    def valid(self):
        sanitized = self.sequence.replace(" ", "")
        if len(sanitized) <= 1:
            return False

        contains_alpha = re.compile(r"\D+")
        if contains_alpha.findall(sanitized):
            return False

        numbers = [int(x) for x in sanitized[::-1]]

        res = 0
        for index, number in enumerate(numbers):
            if (index + 1) % 2 == 0:
                res += doubling(number)
                continue

            res += number

        return res % 10 == 0

File: php/luhn/luhn.php
<?php

function isValid(string $sequence = ""): bool
{
    $sanitized = preg_replace("/\s+/", "", $sequence);
    if (strlen($sanitized) <= 1) {
        return false;
    }

    preg_match("/\D/", $sanitized, $matches);
    if (count($matches) > 0) {
        return false;
    }

    $numbers = array_reverse(
        array_map(function ($number) {
            return intval($number);
        }, str_split($sanitized))
    );

    $res = 0;
    foreach ($numbers as $index => $number) {
        if (($index + 1) % 2 == 0) {
            $res += doubling($number);
            continue;
        }
        $res += $number;
    }

    return $res % 10 == 0;
}

function doubling(int $number): int
{
    $double = $number * 2;
    return $double < 9 ? $double : $double - 9;
}
